src/mlpack/methods/neighbor_search/spill_search_impl.hpp

```diff
@@ -0,0 +1,251 @@
+    const MetricType metric) :
+    neighborSearch(naive, singleMode, epsilon, metric),
+    tau(tau),
+    rho(rho),
+    leafSize(leafSize)
+{
+  if (tau < 0)
+    throw std::invalid_argument("tau must be non-negative");
+  if (rho < 0 || rho > 1)
+    throw std::invalid_argument("rho must be in the range [0,1]");
+}
+
+// Clean memory.
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+~SpillSearch()
+{
+  /* Nothing to do */
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(const MatType& referenceSet)
+{
+  if (Naive())
+    neighborSearch.Train(referenceSet);
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(referenceSet, tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(MatType&& referenceSetIn)
+{
+  if (Naive())
+    neighborSearch.Train(std::move(referenceSetIn));
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(std::move(referenceSetIn), tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(Tree* referenceTree)
+{
+  neighborSearch.Train(referenceTree);
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Search(const MatType& querySet,
+       const size_t k,
+       arma::Mat<size_t>& neighbors,
+       arma::mat& distances)
+{
+  if (Naive() || SingleMode())
+    neighborSearch.Search(querySet, k, neighbors, distances);
+  else
+  {
+    // For Dual Tree Search on SpillTrees, the queryTree must be built with non
+    // overlapping (tau = 0).
+    Tree queryTree(querySet, 0 /* tau */, leafSize, rho);
```

Comment:
@rcurtin Ok, I see what you mean. But, in the last commits, I also define different single and dual tree traversers for defeatist search. So, we can't consider the default traversers for defeatist search, as `NeighborSearch` does.
Because of that I think we should continue considering a different class for `SpillSearch`.
So the main difference with `NeighborSearch` class is:
 + It includes the parameters `LeafSize`, `Tau` and `Rho`, which are specific for SpillTrees.
 + It ensures we build the query tree with non overlappin (tau = 0).
 + It ensures we use the `DefeatistSingleTreeTraverser` and `DefeatistDualTreeTraverser`.

If we want to include all of this in the `NeighborSearch` class, I think it can become a bit complex ....
What do you think?